src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkCustomCommandLine.java

```diff
@@ -336,40 +333,48 @@ private int preprocess(
       }
       List<Object> expandedValues =
           maybeExpandDirectories(artifactExpander, originalValues, pathMapper);
-      List<String> stringValues;
+      List<Object /* String | DerivedArtifact */> values;
       if (mapEach != null) {
-        stringValues = new ArrayList<>(expandedValues.size());
+        values = new ArrayList<>(expandedValues.size());
         applyMapEach(
             mapEach,
             expandedValues,
-            stringValues::add,
+            values::add,
             location,
             artifactExpander,
             pathMapper,
             starlarkSemantics);
       } else {
         int count = expandedValues.size();
-        stringValues = new ArrayList<>(expandedValues.size());
+        values = new ArrayList<>(expandedValues.size());
         for (int i = 0; i < count; ++i) {
-          stringValues.add(expandToCommandLine(expandedValues.get(i), pathMapper, mainRepoMapping));
+          values.add(expandToCommandLine(expandedValues.get(i), mainRepoMapping));
         }
       }
       // It's safe to uniquify at this stage, any transformations after this
       // will ensure continued uniqueness of the values
       if ((features & UNIQUIFY) != 0) {
-        int count = stringValues.size();
+        int count = values.size();
         HashSet<String> seen = Sets.newHashSetWithExpectedSize(count);
         int addIndex = 0;
         for (int i = 0; i < count; ++i) {
-          String val = stringValues.get(i);
-          if (seen.add(val)) {
-            stringValues.set(addIndex++, val);
+          Object /* String | DerivedArtifact */ val = values.get(i);
+          // If the path mapper is a no-op, an artifact behaves just like its (trivially mapped)
+          // exec path string. If the path mapper is not a no-op, mapped paths are always distinct
+          // from unmapped paths. We can thus uniquify based on the mapped exec path string in each
+          // case.
+          String valToDeduplicateBy =
+              val instanceof DerivedArtifact artifact
```

Comment:
This conditional is equivalent to `pathMap(val, pathMapper)`, right?

Also, if you have a better name than `pathMap()` that would be welcome, given that only conditionally applies path mapping. Some people like to call things like this `maybePathMap()`. Feel free to choose your favorite or leave as-is.